eniius/mcstas.py

```diff
@@ -494,13 +523,81 @@ def NXtransformations(self, name):
     def NXcomponent(self, name, order=0):
         # Returns a NXcomponent corresponding to a McStas component.
         comp = self.components[self.indices[name]]
-        mcpars = {p:getattr(comp, p) for p in comp.parameter_names}
+        # pull together component values (or instrument parameter names) for *defined* parameters
+        # -- if a parameter is 'None' at this point, the Component default should be used
+        mcpars = {p: mcstasscript_parameter_name_or_value(getattr(comp, p)) for p in comp.parameter_names if getattr(comp, p) is not None}
         return McStasComp2NX(comp, order, self.NXtransformations(name), **mcpars).nxobj
 
     def NXinstrument(self):
         nxinst = NXinstrument()
+        nxinst['mcstas'] = mcstasscript_instrument_to_nx(self.instrument)
         for order, comp in enumerate(self.components):
             nxinst[comp.name] = self.NXcomponent(comp.name, order)
         return nxinst
 
 
+def _float_int_or_str(s: str):
+    # Parse a string to an integer if possible, or a float if numeric but not integer, or return the string
+    try:
+        v = float(s)
+        return int(v) if v.is_integer() else v
+    except ValueError:
+        return s
+        
+
+def mcstasscript_parameter_name_or_value(parameter):
+    from mcstasscript.helper.mcstas_objects import DeclareVariable
+    from libpyvinyl.Parameters.Parameter import Parameter
+    from ast import literal_eval
+    if isinstance(parameter, (DeclareVariable, Parameter)):
+        return parameter.name
+    return _float_int_or_str(parameter)
+
+
+def mcstasscript_parameter_to_nexus(parameter, index=None):
+    from mcstasscript.helper.mcstas_objects import DeclareVariable
+    from libpyvinyl.Parameters.Parameter import Parameter
+    if isinstance(parameter, DeclareVariable):
+        fields = ('type', 'name','value', 'comment')
+    elif isinstance(parameter, Parameter):
+        fields = ('type', 'name','value', 'unit', 'comment')
+    else:
+        # str, float, int, list, ... can all be handled directly
+        # (only str *should* be encountered, for %include library lines)
+        name = '__item__' if index is None else f'__item{index}__'
+        return parameter, name
+
+    attributes = {f: getattr(parameter, f) for f in fields if hasattr(parameter, f)}
+    # filter out empty-string (or anything equivalent to False) valued attributes since they cause problems :/
+    attributes = {k: v for k, v in attributes.items() if v}
+    out = NXfield(**attributes)
+    return out, parameter.name
+
+
+def mcstasscript_parameter_list_to_nx(name, parameters):
+    nx = NXcollection(name=name)
+    for i, p in enumerate(parameters):
+        nxp, name = mcstasscript_parameter_to_nexus(p, index=i)
+        nx[name] = nxp
+    return nx
+
+
+def mcstasscript_instrument_to_nx(instrument):
+    # Follow the McStasScript instrument writing logic:
+    mcinst = NXcollection(name=instrument.name, version='3.3')
```

Comment:
Only the major version at the moment, which might not be safe-enough since new features can be added in minor releases. Maybe I can convince @mads-bertelsen to change `McStas_instr.mccode_version` to be the full version number of the mcstas binary McStasScript is using?